### ISSUE TITLE
Bad lint warning

### ISSUE THREAD

**Issue by Inky-developer** [rust-lang/rust-clippy#5778]
I tried this code:

```rust
// old_data is some Option<T>
let mut should_write = old_data.is_none();
if old_data.map_or(false, |old_data| old_data != watcher) {
    // do stuff
    should_write = true;
}
```

I expected to see this happen: No clippy lint / other clippy lint

Instead, this happened: warning `if _ { .. } else { .. } is an expression` appeared. Clippy advises to do this: `help: it is more idiomatic to write: 'let <mut> should_write = if old_data.map_or(false, |old_data| old_data != watcher) { ..; true } else { old_data.is_none() };'`
However, this code will not compile as `old_data.is_none()` would be evaluated after `old_data.map_or(...)`

### Meta

- `cargo clippy -V`: clippy 0.0.212 (bb37a0f9 2020-06-16)
- `rustc -Vv`:
  ```
    rustc 1.44.1 (c7087fe00 2020-06-17)
    binary: rustc
    commit-hash: c7087fe00d2ba919df1d813c040a5d47e43b0fe7
    commit-date: 2020-06-17
    host: x86_64-pc-windows-msvc
    release: 1.44.1 
    LLVM version: 9.0
  ```

**Comment by montrivo** [comment]
This [example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=cc8c9983f4ff29d03d9a7bda6b46a7fc) based on the issue description does not trigger the lint. 

@Inky-developer could you prepare an example in the playground that reproduces the issue?

**Comment by Inky-developer** [comment]
Strange, this error does not pop up anymore, even in the original code. I will assume that this was fixed and close the issue for now.
